Lesson eleven

DADUSHKA opened the pull request.

Vicstorovich committed:
UserDevise

Vicstorovich committed:
Mailer

Vicstorovich committed:
Lesson eleven is ready

BubuntuClu commented on app/models/test.rb:
```diff
@@ -1,8 +1,7 @@
 class Test < ApplicationRecord
-  scope :by_category, ->(category_name) do
+  scope :by_category, lambda { |category_name|
```
обычно когда `{ }` используют, это значит что выражение будет в одну строку, иначе `do... end` ставят

BubuntuClu commented on app/models/test_passage.rb:
```diff
@@ -15,21 +15,21 @@ def accept!(answer_ids)
-    else
-      self.question = test.questions.first
-    end
+    self.question = if !question_id.nil?
```
```
self.question = 
  if !question_id.nil?
    test.questions.order(:id).where('id > ?', question.id).first
  else
    test.questions.first
  end                     
```
так тоже допустимо, и выглядит вполне опрятно

BubuntuClu commented on app/views/admin/answers/_answer.html.erb:
```diff
@@ -0,0 +1,7 @@
+<li>
+<%= answer.body %><br>
```
отступ

BubuntuClu commented on app/views/layouts/application.html.erb:
```diff
@@ -12,6 +12,6 @@
         <%= render "shared/flash_message" %>
         <%= yield %>
-        <%= debug(flash) if Rails.env.development? %>
+        <%#= debug(params) if Rails.env.development? %>
```
удаляй это сразу

BubuntuClu commented on app/views/shared/_menu.html.erb:
```diff
@@ -1,15 +1,15 @@
-    <li><%= link_to 'Log out', session_path(session),
+    <li></li>
+    <% if user_signed_in?  %>
+    <li>Welcome <%= "#{current_user.first_name}""  ""#{current_user.last_name}" %></li>
```
отступ

BubuntuClu replied:
можешь метод в моделе пользователя сделать, который тебе вернет нужную строку с именем

BubuntuClu commented on app/views/tests/index.html.erb:
```diff
@@ -1,7 +1,9 @@
   <div id="sidebar">
-    <%= link_to 'New tests', new_test_path %>
+<% if current_user.admin_status? %>
+<%= link_to 'AdminPanel',  admin_tests_path %>
```
отступ

BubuntuClu commented on db/seeds.rb:
```diff
@@ -8,90 +8,107 @@
-     password_digest: 'q3we9rty' }]
-)
+
+# users = User.create!(
```
а что тут такое случилось? почему все убрал?

psylone commented on app/controllers/admin/admin_controller.rb:
```diff
@@ -0,0 +1,12 @@
+  protected
+
+  def check_admin
+    redirect_to root_path, alert: 'You are not authorized to access this page' unless current_user.admin_status?
```
Лучше просто `current_user.admin?`

psylone commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -0,0 +1,58 @@
+  private
+
+  def test_params
+    params.require(:test).permit(:title, :level, :author_id, :category_id)
```
Используй метод ассоциации при создании объекта теста, тогда `:author_id` здесь не будет нужен.

psylone commented on db/migrate/20190110213655_add_first_name_last_name_to_users.rb:
```diff
@@ -0,0 +1,8 @@
+class AddFirstNameLastNameToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :first_name, :string, null: false, default: ''
+    add_column :users, :last_name, :string, null: false, default: ''
```
Лучше оставить `nil` по умолчанию.

DADUSHKA replied:
В скринкасте сказано, что так Devise требует.

psylone replied:
Согласен, учитывая `null: false`. Хотя на самом деле, это зависит от конкретной ситуации.

psylone commented on db/migrate/20190110213655_add_first_name_last_name_to_users.rb:
```diff
@@ -0,0 +1,8 @@
+    add_column :users, :first_name, :string, null: false, default: ''
+    add_column :users, :last_name, :string, null: false, default: ''
+
+    remove_column(:users, :name)
```
Это не сработает для `change`, потому что Rails не будет знать какой тип был у атрибута если откатывать миграцию.

psylone replied:
Не вполне ясно что здесь теперь происходит: переопределяется метод `change_column`?

psylone commented:
Что с усложнённым заданием?

BubuntuClu commented:
> Что с усложнённым заданием?

[вот ведь](https://github.com/DADUSHKA/web_application/pull/13/files#diff-d5a7c21e9ed066de2a48c858ecd8d713R6)

Vicstorovich committed:
Lesson eleven misstake 1

DADUSHKA committed:
Как-то так

psylone commented on app/controllers/sessions_controller.rb:
```diff
@@ -0,0 +1,13 @@
+
+  def create
+    self.resource = warden.authenticate!(auth_options)
+    flash[:notice] = "Welcome back, #{current_user.first_name}!" if is_flashing_format?
```
Достаточно установить флэш сообщение и вызвать `super`.

psylone commented on app/helpers/users_helper.rb:
```diff
@@ -1,2 +1,9 @@
 module UsersHelper
+  def first_name
```
В таких хэлперах особого смысла нет.

psylone commented on app/controllers/admin/tests_controller.rb:
```diff
@@ -17,6 +17,7 @@ def edit; end
 
   def create
     @test = Test.new(test_params)
+    @test.author_id = current_user.id
```
Что-то не получилось с использованием метода ассоциации?

DADUSHKA replied:
додумался так, а как еще можно?

psylone replied:
`current_user.author_tests.new(test_params)`.

DADUSHKA committed:
Update 20190110213655_add_first_name_last_name_to_users.rb

DADUSHKA committed:
Update sessions_controller.rb

DADUSHKA reviewed (commented):
это чтобы не писать так - <li>Welcome <%= "#{current_user.first_name}""  ""#{current_user.last_name}" %></li>

DADUSHKA commented:
.

DADUSHKA committed:
Update users_helper.rb

это чтобы не писать так - <li>Welcome <%= "#{current_user.first_name}""  ""#{current_user.last_name}" %></li>
а так - <li>Welcome <%= first_name %> <%= last_name}" %></li>
